File: Julia/search/jump_search.jl
"""Julia program to implement Jump Search algorithm.
Jump Search is a searching algorithm similar to binary search, where a sorted array is searched. 
In this algorithm, only fewer comparisons are done, by jumping ahead some elements while traversal
"""


function jump_search(arr, num, ele)

    # Size by which the traversing index should be jumped is calculated.
    jump = floor(Int, sqrt(num))
    step = jump
    temp = 1
    
    """ The entire sorted array is divided into blocks of sqrt(n) size and the block just after which the
        required element is present is being identified. and the previous block starting index is stored in temp""" 
    while (arr[min(step, num)] < ele)
        temp = step
        step = step + jump
        if (temp >= num)
            return false
        end
    end

    # The identified block, is searched via linea search for the required element.
    while (arr[temp] < ele)
        temp = temp + 1
        if (arr[temp] == ele)
            return true
        end
        # Check if the end of the array or end of the block is reached 
        if (temp == min(step, num))
            return false
        end
    end
    if (arr[temp] == ele)
        return true
    else
        return false
    end
end


print("How many numbers are present in the array? ")
n = readline()
n = parse(Int, n)
if (n <= 0)
    println("Array is Empty!!!")
    exit()
end
arr = Int[]
print("Enter the numbers: ")
arr = [parse(Int, num) for num in split(readline())] 
print("Which number do you want to search in the array? ")
ele = readline()
ele = parse(Int, ele)
# Sort the array in ascending order
arr = sort(arr)
res = jump_search(arr, n, ele)
if (res == 0)
    print("The number $ele is not present in the array")
else
    print("The number $ele is present in the array.")
end



"""
Time Complexity - O(n^(0.5)), where 'n' is the size of the array
Space Complexity - O(n)

SAMPLE INPUT AND OUTPUT

SAMPLE I

How many numbers are present in the array? 5 
Enter the numbers: 1 2 3 4 5
Which number do you want to search in the array? 6
The number 6 is not present in the array

SAMPLE II

How many numbers are present in the array? 3
Enter the numbers: 3 1 2
Which number do you want to search in the array? 2
The number 2 is present in the array.

"""

File: Dart/search/Jump_Search.dart
/*
JUMP SEARCH IN DART
Jump search is a sorting algorithm suited for ordered lists. 
It creates a block and tries to find the element in that block. 
If the item is not in the block, it shifts the entire block. 
The block size chosen is √n, where n is the size of the array.
*/

import 'dart:io';
import 'dart:math';

int jumpSearch(List arr, int len, int key) {
  int begin = 0;
  // Setting the size of each block as squareroot of length of array
  int block_size = sqrt(len).floor();

  // Finding the block in which the key may potentially lie
  while (arr[block_size] <= key && block_size < len) {
    // setting the beginning index for the next block
    begin = block_size;
    // setting the ending index for the next block
    block_size += sqrt(len).floor();

    // if the block_size index lies beyond the length of array
    // indicating the key may or may not lie in last block
    // if it lies then linear search will find it and if not
    // return -1
    if (block_size > len - 1) {
      block_size = len;
      break;
    }
  }

  // performing linear search in the block in which the key may lie
  for (int ind = begin; ind < block_size; ind += 1) {
    if (arr[ind] == key) return ind;
  }
  return -1;
}

void main() {
  var n, ele, key;
  print('Enter the number of elements (n): ');
  n = stdin.readLineSync();
  n = int.parse(n);

  List array = new List.filled(n, null, growable: false);

  for (int i = 0; i < n; i++) {
    print('Enter value for element $i: ');
    ele = stdin.readLineSync();
    ele = int.parse(ele);
    array[i] = ele;
  }

  print('Enter the element to be searched (key): ');
  key = stdin.readLineSync();
  key = int.parse(key);

  int index = jumpSearch(array, n, key);

  if (index >= 0)
    print('$key was found at index $index');
  else
    print('$key is not present in the given array');
}

/*
TIME COMPLEXITY: O(√n)
SPACE COMPLEXITY: O(1)

SAMPLE INPUT/OUTPUT:
Enter the number of elements (n): 
7
Enter value for element 0: 
11
Enter value for element 1:
12
Enter value for element 2:
33
Enter value for element 3:
44
Enter value for element 4:
55
Enter value for element 5:
66
Enter value for element 6:
77
Enter the element to be searched (key):
55
55 was found at index 4
*/
